### ISSUE TITLE
Programs: Add "Interactive Mode" to delay alerted program and let user to allow/block it.

### ISSUE THREAD

**Issue by tnodir** [tnodir/fort#30]

**Comment by tnodir** [comment]
Yes, this is. Ok, let me add some basic functionality…

**Comment by wealstarr** [comment]
> Thanks! Something similar to [Netlimiter's prompt](https://netlimiter.com/docs/basic-concepts/blocker/pending-requests) would probably work for most use cases

Nope, that doesn't work. It only asks for applications that you have added in the "ask" category. Who knows how many programs and services are connecting from behind your back. 

It should be like xvirus firewall, where any app that tries to connect invokes a notification and ability to record the answer in settings permanently.

**Comment by ahdung** [comment]
+1, Very need this feature.

**Comment by Littleweisheit** [comment]
This function is very important, please put this as your first priority

**Comment by tnodir** [comment]
> This function is very important

Why?

Most firewalls don't pause connections, but just show notification about blocked ones: Windows Firewall Control (WFC) by BiniSoft.org, Simplewall, GlassWire.

Portmaster, NetLimiter pause connections, because they have drivers.

(I'm going to work on this feature on winter holidays. And now I'm doing other easy ones.)

**Comment by tnodir** [comment]
> I am using Simplewall,It will ask me for allow or block,before that Simplewall will block connection.

1. Simplewall provides rules to Windows Filtering Platform (WFP) and does not block connections itself
2. Simplewall shows notification about blocked connection, because it can not pause connections

Fort also shows notification about blocked connections. I don't see a difference.

**Comment by Littleweisheit** [comment]
> > I am using Simplewall,It will ask me for allow or block,before that Simplewall will block connection.
> 
>     1. Simplewall provides rules to Windows Filtering Platform (WFP) and does not block connections itself
> 
>     2. Simplewall shows notification about blocked connection, because it can not pause connections
> 
> 
> Fort also shows notification about blocked connection. So, I don't see a difference.

When “Fort” support "Ask to Connect"，The difference will be small.
Simplewall not support path like: d:softwate/*/abc.exe the 
**_"*"_** in path usually  version number

**Comment by ahdung** [comment]
@tnodir But simplewall can **Allow** in the dialog, user don't care if it is `pause` or `block` the connection I think, we just need a dialog which provided a `Allow` and `Block` choices.

**Comment by ahdung** [comment]
I know the difference, I want to saying it doesn't matter, retry is fine, not too much inconvenience, user accepts retry, not accepts creating rules manually, that's the point, so **Allow** is not `continue` connection, it's just `creating` a rule automatically.

**Comment by danielmmmm** [comment]
For users who value privacy and security more than convenience, even blocking a connection would be preferable to allowing a connection, until the user choses to block or allow on a pop-up dialogue :-)
A failed installation can be retried. A data leak is final and cannot be reversed.

**Comment by SimoLRepo** [comment]
> > This function is very important
> 
> Why?
> 
> Most firewalls don't pause connections, but just show notification about blocked ones: Windows Firewall Control (WFC) by BiniSoft.org, Simplewall, GlassWire.
> 
> Portmaster, NetLimiter pause connections, because they have drivers.
> 
> (I'm going to work on this feature on winter holidays. And now I'm doing other easy ones.)

I also join the supporters of this feature, by the way essential for any self-respecting firewall. At first glance this firewall already has pretty much everything you need, except a popup notification unfortunately. The absence of such functionality prevents me from using it as my main firewall.

The <b>BiniSoft Windows Firewall Control</b> actually temporarily blocks the connection and then alerts the user whether to allow it or not, personally tested with installers that required online content downloads.

**Comment by tnodir** [comment]
> The BiniSoft Windows Firewall Control actually temporarily blocks

From WFC’s [manual](https://www.binisoft.org/pdf/guides/Malwarebytes-WFC-User-Guide.pdf):
“The notifications are displayed for blocked connections, not for paused”

**Comment by ahdung** [comment]
@tnodir Please don't care if it's a real "pause", we just want a button to create rules with one click on demand.

**Comment by Littleweisheit** [comment]
> @tnodir Please don't care if it's a real "pause", we just want a button to create rules with one click on demand, this has worked well for many years on simplewall, the only problem with it for me is that it doesn't support wildcard fuzzy matching, so if your two tools combined that would be perfect.

yeah,for now I just care "Ask to allow connect"

**Comment by SimoLRepo** [comment]
> > The BiniSoft Windows Firewall Control actually temporarily blocks
> 
> From WFC’s [manual](https://www.binisoft.org/pdf/guides/Malwarebytes-WFC-User-Guide.pdf): “The notifications are displayed for blocked connections, not for paused”

WFC evidently does not consider "paused" (TIME_WAIT) connections simply because they have already been passed and allowed by the user. The connections that remain in this state are usually legitimate system processes, blocking them would not make much sense unless you set a "paranoid" mode.

However, I have read some threads on MalwareTips that report instabilities in this software. Until it's stable I won't be able to use it.

**Comment by tnodir** [comment]
> I have read some threads on MalwareTips that report instabilities in this software.

@SimoLRepo Do you mean instabilities in Fort Firewall? Then they were fixed already.

**Comment by SimoLRepo** [comment]
> Do you mean instabilities in Fort Firewall? Then they were fixed already.

Okay, thanks for the clarification

**Comment by tnodir** [comment]
Partially implemented in [v3.11.0](https://github.com/tnodir/fort/releases/tag/v3.11.0).
